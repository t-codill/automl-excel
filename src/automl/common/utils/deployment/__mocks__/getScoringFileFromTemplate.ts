import { RunHistoryAPIsModels } from "@vienna/runhistory";

export const getScoringFileFromTemplate = (
    parentRun: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto | undefined,
    runId: string | undefined,
): string | undefined => {
    if (!runId || !parentRun) {
        return "";
    }

    return `mock scoring file content for runId: ${runId}, run: ${parentRun.runId}`;
};
