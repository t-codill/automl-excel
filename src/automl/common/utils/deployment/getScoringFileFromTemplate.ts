import { RunHistoryAPIsModels } from "@vienna/runhistory";
import { getModelNameFromRunId } from "../getModelNameFromRunId";
import { safeParseJson } from "../safeParseJson";

const schemaTemplate = `
from inference_schema.schema_decorators import input_schema, output_schema
from inference_schema.parameter_types.numpy_parameter_type import NumpyParameterType

input_sample = np.array([[##input##]])
output_sample = np.array(##output##])
`;

const scoringTemplate = `
import pickle
import json
import azureml.train.automl
import pandas as pd
from sklearn.externals import joblib
from azureml.core.model import Model

def init():
    global model
    model_path = Model.get_model_path(model_name='##modelName##')
    model = joblib.load(model_path)

##schemaDecorators##
def run(rawdata):
    try:
        data = json.loads(rawdata)['data']
        df = pd.DataFrame(data,columns=##featureColumns##)
        result = model.predict(df)
    except Exception as e:
        result = str(e)
        return json.dumps({"error": result})
    return json.dumps({"result": result.tolist()})
`;

export const getScoringFileFromTemplate = (
    parentRun: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto | undefined,
    runId: string | undefined,
    inputSample?: any,
    outputSample?: any
) => {
    const dataPrepJson = parentRun && parentRun.properties && parentRun.properties.DataPrepJsonString;
    if (!dataPrepJson || !runId) {
        return undefined;
    }

    const dataPrep = safeParseJson(safeParseJson(`"${dataPrepJson}"`));
    if (!dataPrep || !dataPrep.features) {
        return undefined;
    }

    const featureColumns = JSON.stringify(dataPrep.features);
    const modelName = getModelNameFromRunId(runId);

    let scoringScript = scoringTemplate
    .replace("##modelName##", modelName)
    .replace("##featureColumns##", featureColumns);

    if(inputSample !== undefined && outputSample != undefined){
        scoringScript = scoringScript.replace("##schemaDecorators##", schemaTemplate);
    }else{
        scoringScript = scoringScript.replace("##schemaDecorators##", "");
    }

    return scoringScript;
};
