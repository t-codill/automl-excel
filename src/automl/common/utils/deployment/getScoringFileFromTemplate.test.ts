import { getScoringFileFromTemplate } from "./getScoringFileFromTemplate";

describe("getScoringFileFromTemplate", () => {
    it("should return undefined if run Id is undefined", () => {
        expect(getScoringFileFromTemplate({}, undefined))
            .toBeUndefined();
    });

    it("should return undefined if run is undefined", () => {
        expect(getScoringFileFromTemplate(undefined, "foo"))
            .toBeUndefined();
    });

    it("should return undefined if run properties is undefined", () => {
        expect(getScoringFileFromTemplate({ properties: undefined }, "foo"))
            .toBeUndefined();
    });

    it("should return undefined if DataPrepJsonString is empty", () => {
        expect(getScoringFileFromTemplate({ properties: { DataPrepJsonString: "" } }, "foo"))
            .toBeUndefined();
    });

    it("should return undefined if DataPrepJsonString is invalid", () => {
        expect(getScoringFileFromTemplate({ properties: { DataPrepJsonString: "invalid" } }, "foo"))
            .toBeUndefined();
    });

    it("should return undefined if DataPrepJsonString features is missing", () => {
        expect(getScoringFileFromTemplate({ properties: { DataPrepJsonString: "{}" } }, "foo"))
            .toBeUndefined();
    });

    it("should return valid scoring file with valid DataPrepJsonString", () => {
        expect(getScoringFileFromTemplate({
            properties: {
                DataPrepJsonString: "{\\\"features\\\":[\\\"timeStamp\\\",\\\"precip\\\",\\\"temp\\\"]}"
            }
        }, "foo"))
            .toMatchSnapshot();
    });
});
