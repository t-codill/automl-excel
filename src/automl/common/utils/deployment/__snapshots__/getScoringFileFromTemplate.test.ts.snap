// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getScoringFileFromTemplate should return valid scoring file with valid DataPrepJsonString 1`] = `
"
import pickle
import json
import azureml.train.automl
import pandas as pd
from sklearn.externals import joblib
from azureml.core.model import Model

def init():
    global model
    model_path = Model.get_model_path(model_name='foo')
    model = joblib.load(model_path)

def run(rawdata):
    try:
        data = json.loads(rawdata)['data']
        df = pd.DataFrame(data,columns=[\\"timeStamp\\",\\"precip\\",\\"temp\\"])
        result = model.predict(df)
    except Exception as e:
        result = str(e)
        return json.dumps({\\"error\\": result})
    return json.dumps({\\"result\\": result.tolist()})
"
`;
